src/motherduck_destination_server.cpp

```diff
@@ -0,0 +1,448 @@
+        auto maybe_compressed_input_stream = arrow::io::CompressedInputStream::Make(codec.get(), buffer_reader);
+        if (!maybe_compressed_input_stream.ok()) {
+            throw std::runtime_error("Could not input stream from compressed buffer: " +
+                                     maybe_compressed_input_stream.status().message());
+        }
+        auto compressed_input_stream = std::move(maybe_compressed_input_stream.ValueOrDie());
+
+        auto read_options = arrow::csv::ReadOptions::Defaults();
+        auto parse_options = arrow::csv::ParseOptions::Defaults();
+        auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+        auto maybe_table_reader = arrow::csv::TableReader::Make(arrow::io::default_io_context(),
+                                                                std::move(compressed_input_stream), read_options,
+                                                                parse_options, convert_options);
+        if (!maybe_table_reader.ok()) {
+            throw std::runtime_error("Could not create table reader: " + maybe_table_reader.status().message());
+        }
+        auto table_reader = std::move(maybe_table_reader.ValueOrDie());
+
+        auto maybe_table = table_reader->Read();
+        if (!maybe_table.ok()) {
+            throw std::runtime_error("Could not read CSV" + maybe_table.status().message());
+        }
+        auto table = std::move(maybe_table.ValueOrDie());
+
+        return table;
+    }
+
+    const std::string
+    get_encryption_key(const std::string &filename, const google::protobuf::Map<std::string, std::string> &keys) {
+        auto encryption_key_it = keys.find(filename);
+
+        if (encryption_key_it == keys.end()) {
+            throw std::invalid_argument("Missing encryption key for " + filename);
+        }
+
+        return encryption_key_it->second;
+    }
+
+    std::vector<std::string>
+    get_primary_keys(const google::protobuf::RepeatedPtrField<fivetran_sdk::Column> &columns) const {
+        std::vector<std::string> primary_keys;
+        for (auto &col: columns) {
+            if (col.primary_key()) {
+                primary_keys.push_back(col.name());
+            }
+        }
+        return primary_keys;
+    }
+
+    void process_file(Connection &con,
+                      const std::string &filename,
+                      const std::string &decryption_key,
+                      const std::function<void(std::string view_name)> &process_view) {
+        auto table = ReadCsv(filename, decryption_key);
+
+        auto batch_reader = std::make_shared<arrow::TableBatchReader>(*table);
+        ArrowArrayStream arrow_array_stream;
+        auto status = arrow::ExportRecordBatchReader(batch_reader, &arrow_array_stream);
+        if (!status.ok()) {
+            throw std::runtime_error("Could not convert Arrow batch reader to an array stream: " + status.message());
+        }
+
+        duckdb_connection c_con = (duckdb_connection) &con;
```

Comment:
Is the casting always going to succeed?
Also we should use `static_cast` rather than old-style C cast.

Reply:
I think so? It's how [duckdb does it](https://github.com/duckdb/duckdb/blob/11e18680513b71520bb8e8b109a59772cd7d7513/src/main/capi/duckdb-c.cpp#L61)

Reply:
I used reinterpret_cast.